cpp/src/parquet/encryption/encryption_internal.cc

```diff
@@ -438,56 +460,105 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
   return decryptor;
 }
 
-int AesDecryptor::CiphertextSizeDelta() { return impl_->ciphertext_size_delta(); }
-
-int AesDecryptor::AesDecryptorImpl::GcmDecrypt(const uint8_t* ciphertext,
-                                               int ciphertext_len, const uint8_t* key,
-                                               int key_len, const uint8_t* aad,
-                                               int aad_len, uint8_t* plaintext) {
-  int len;
-  int plaintext_len;
+int AesDecryptor::PlaintextLength(int ciphertext_len) const {
+  return impl_->PlaintextLength(ciphertext_len);
+}
 
-  uint8_t tag[kGcmTagLength];
-  memset(tag, 0, kGcmTagLength);
-  uint8_t nonce[kNonceLength];
-  memset(nonce, 0, kNonceLength);
+int AesDecryptor::CiphertextLength(int plaintext_len) const {
+  return impl_->CiphertextLength(plaintext_len);
+}
 
+int AesDecryptor::AesDecryptorImpl::GetCiphertextLength(
+    span<const uint8_t> ciphertext) const {
   if (length_buffer_length_ > 0) {
+    // Note: length_buffer_length_ must be either 0 or kBufferSizeLength
+    if (ciphertext.size() < static_cast<size_t>(kBufferSizeLength)) {
+      std::stringstream ss;
+      ss << "Ciphertext buffer length " << ciphertext.size()
+         << " is insufficient to read the ciphertext length."
+         << " At least " << kBufferSizeLength << " bytes are required.";
+      throw ParquetException(ss.str());
+    }
+
     // Extract ciphertext length
     int written_ciphertext_len = ((ciphertext[3] & 0xff) << 24) |
                                  ((ciphertext[2] & 0xff) << 16) |
                                  ((ciphertext[1] & 0xff) << 8) | ((ciphertext[0] & 0xff));
 
-    if (ciphertext_len > 0 &&
-        ciphertext_len != (written_ciphertext_len + length_buffer_length_)) {
-      throw ParquetException("Wrong ciphertext length");
+    if (written_ciphertext_len < 0) {
```

Comment:
I wonder if the compiler won't elide this, as `((ciphertext[3] & 0xff) << 24)` becoming negative implies signed integer overflow which is undefined behavior. @felipecrv What do you think?

Reply:
This check is weird. `written_ciphertext_len` should be `uint32_t`, then checking that `written_ciphertext_len` is not greater than `std::numeric_limits<int32_t>::max()` is enough. In other words: the encoded value is never negative, but it can be too big.

Reply:
`ciphertext[3] & 0xff` is weird as well, since `ciphertext[3]` is a uint8, thus in [0,255] already.
Perhaps the proper check should simply be `ciphertext[3] < 0x80u`?

Reply:
Hmm yeah the `& 0xff`s are all redundant, maybe it ended up written this way to be similar to the code for writing the length. Rather than just checking the length isn't negative, we should also check that `written_ciphertext_len + length_buffer_length_` doesn't overflow, so I think it would probably be simplest to read as `uint32_t`.

Reply:
I've changed this check as part of #43195, in this commit: https://github.com/apache/arrow/pull/43195/commits/b81cafe1a74cd821fde4d2e58481359149eaf233